### PR TITLE
libdeflate: Update to v1.17

### PR DESCRIPTION
**Pull request by oliv3r**
CMake depends on (libdeflate-)gunzip, libdeflate depends on Cmake, so we can't win.

Luckily libdeflate is _very_ easy to build, without any build system, so lets just manually compile it and be done with it.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

### CODE REVIEW

File: tools/Makefile [hunk 1]
```diff
@@ -89,7 +89,7 @@ $(curdir)/genext2fs/compile := $(curdir)/libtool/compile
 $(curdir)/gengetopt/compile := $(curdir)/libtool/compile
 $(curdir)/gmp/compile := $(curdir)/libtool/compile
 $(curdir)/isl/compile := $(curdir)/gmp/compile
-$(curdir)/libdeflate/compile := $(curdir)/cmake/compile
+$(curdir)/libdeflate/compile := $(curdir)/libdeflate/compile
```
Comment (neheb) [inline tools/Makefile:92]: uhhh what?

File: tools/libdeflate/Makefile [hunk 2]
```diff
@@ -16,7 +16,9 @@ PKG_SOURCE_VERSION:=v$(PKG_VERSION)
 PKG_MIRROR_HASH:=122feff4543541b547dc89e832adf262c81911ae1acbccdc591f0353a85b600a
 
 include $(INCLUDE_DIR)/host-build.mk
-include $(INCLUDE_DIR)/cmake.mk
+
+define Host/Compile
+	$(CC) lib/*{,/*}.c programs/{gzip,prog_util,tgetopt}.c -o libdeflate-gzip
```
Comment (neheb) [inline tools/libdeflate/Makefile:21]: HOST_CFLAGS should also be passed.

File: tools/Makefile (hunk 1 above)
Comment (oliv3r) [reply tools/Makefile:92]: I have no idea :D but would appreciate some help here. I was looking at the other stuff above. basically, we only depend on gccc ...

Comment (neheb) [reply tools/Makefile:92]: delete this line. libdeflate is not dependent on anything after this change.

**Comment by oliv3r** [comment]
@neheb btw, how can I force a rebuild of a tool? It keeps telling me 'everything is up to date'. `make tools/libdeflate` with clean, compile, anything or everything, `make oldconfig` nothing yields a rebuild ... (i tried searching for libdeflate stamps and artifacts, but found nothing to delete anymore. ...

**Comment by robimarko** [comment]
You tried changing the `PKG_RELEASE`?

**Comment by neheb** [comment]
I think it's because this package is using Host/Clean instead of Host/Uninstall

**Comment by Ansuel** [comment]
Oh wow it can simply be compiled with good old gcc raw command? Nice!

**Comment by oliv3r** [comment]
> You tried changing the `PKG_RELEASE`?

Part of the Merge Request already :) but no :( doesn't work.

**Comment by oliv3r** [comment]
> I think it's because this package is using Host/Clean instead of Host/Uninstall
> 
> edit: yeah, need to call Host/Clean/Default in that section.

So this goes beyond my knowledge (and desire to know) about the OpenWRT build system :p however, no, doesn't seem to make a difference. I tried several combinations, Host/Uninstall, Host/Clean; adding the call (I did push that bit as I saw other tools having it). But a) I have no idea how all of this magic works, and b) I can't force a rebuild to check if it actually compiles. So feel free to merge it on good faith :) or test it somehow. I can't seem to :(

**Comment by neheb** [comment]
I'll fix it in the other PR.

**Comment by oliv3r** [comment]
@neheb thank you for fixing it in your MR, i'll close this and refer to yours!

**Comment by oliv3r** [comment]
So I figured things here, which due to a version bump makes building even easier.

**Comment by oliv3r** [comment]
@neheb should be all good now I hope :)

File: tools/libdeflate/Makefile [hunk 3]
```diff
@@ -8,24 +8,32 @@ include $(TOPDIR)/rules.mk
...
 PKG_MIRROR_HASH:=ee5790cf3140aa6a2e0f0c400d4b32539f13cb270e9357135c51927ba3784dc7
 
 include $(INCLUDE_DIR)/host-build.mk
-include $(INCLUDE_DIR)/cmake.mk
+
+define Host/Compile
+	$(HOSTCC_NOCACHE) $(HOST_CFLAGS) $(HOST_LDFLAGS) \
+	$(HOST_BUILD_DIR)/lib/*{,/*}.c \
+	$(HOST_BUILD_DIR)/programs/{gzip,prog_util,tgetopt}.c \
+	-I$(HOST_BUILD_DIR) \
```
Comment (neheb) [inline tools/libdeflate/Makefile:24]: Remove

**Comment by neheb** [comment]
Change it to libdeflate: update to 1.17

File: tools/libdeflate/Makefile (hunk 3 above)
Comment (oliv3r) [reply tools/libdeflate/Makefile:24]: Everything? or just 'programs'?

Comment (neheb) [reply tools/libdeflate/Makefile:24]: The -I line is not needed.

Comment (oliv3r) [reply tools/libdeflate/Makefile:24]: ah, ok fair nuff!

File: tools/Makefile [hunk 4]
```diff
@@ -136,8 +134,9 @@ $(foreach tool, $(tools-y), $(if $(wildcard $(curdir)/$(tool)/patches),$(eval $(
 
 $(foreach tool, $(filter-out xz,$(tools-y)), $(eval $(curdir)/$(tool)/compile += $(curdir)/xz/compile))
 
-# make any tool depend on tar, xz and patch to ensure that archives can be unpacked and patched properly
-tools-core := tar xz patch
+# make any tool depend on libdeflate, patch, tar, and xz to ensure that archives
+# can be unpacked and patched properly
+tools-core := libdeflate patch tar xz
```
Comment (neheb) [inline tools/Makefile:139]: Doesn't belong in this commit. You're not changing the unpack routine.

**Comment by neheb** [comment]
Change it to libdeflate: update to 1.17 and keep second commit title lowercase.

**Comment by oliv3r** [comment]
> Change it to libdeflate: update to 1.17 and keep second commit title lowercase.

"Officially" Linux and in many git guidelines, you are supposed to start your subject with a capital. I'll make it consistent.

File: tools/Makefile (hunk 4 above)
Comment (oliv3r) [reply tools/Makefile:139]: Are you telling me order matters? You put libdeflate as part of the tools-core, so I took that over, and fixed sorting as part of this.

Comment (Ansuel) [reply tools/Makefile:139]: should we change this list like the tools one (one tool per line) for consistency and better tracking?

Comment (neheb) [reply tools/Makefile]: no. this PR doesn't force unpacking with libdeflate. why would it be moved to tools-core?

File: tools/libdeflate/Makefile [hunk 5]
```diff
@@ -23,7 +23,7 @@ define Host/Install
 	$(LN) libdeflate-gzip $(STAGING_DIR_HOST)/bin/libdeflate-gunzip
 endef
 
-define Host/Clean
+define Host/Uninstall
 	rm -f $(STAGING_DIR_HOST)/bin/libdeflate-gzip
 	rm -f $(STAGING_DIR_HOST)/bin/libdeflate-gunzip
```
Comment (neheb) [inline tools/libdeflate/Makefile]: you're missing Uninstall/Default

Comment (neheb) [reply tools/libdeflate/Makefile]: nvm. next commit adds it.

File: tools/Makefile (hunk 4 above)
Comment (oliv3r) [reply tools/Makefile]: That's something I have no idea for; but I took this from your set of changes @neheb

**Comment by Ansuel** [comment]
@oliv3r also can you change `(actually fix the PR so that it works)` to [ ... ] ? 

With the 2 requeste change, I will merge this and then I will merge https://github.com/openwrt/openwrt/pull/11717 rebased

**Comment by oliv3r** [comment]
> @oliv3r also can you change `(actually fix the PR so that it works)` to [ ... ] ?
> 
> With the 2 requeste change, I will merge this and then I will merge #11717 rebased

Done! Lets get this puppy in!

**Comment by Ansuel** [comment]
@oliv3r missing separator error

**Comment by Ansuel** [comment]
There are some rebase comment left

File: tools/Makefile [hunk 6]
```diff
@@ -98,8 +97,11 @@ $(curdir)/genext2fs/compile := $(curdir)/libtool/compile
 $(curdir)/gengetopt/compile := $(curdir)/libtool/compile
 $(curdir)/gmp/compile := $(curdir)/libtool/compile
 $(curdir)/isl/compile := $(curdir)/gmp/compile
+<<<<<<< HEAD
```
Comment (Ansuel) [inline tools/Makefile:100]: here

Comment (oliv3r) [reply tools/Makefile]: yeah, no idea how that snook in ... 0bviously during rebase :p

**Comment by Ansuel** [comment]
@oliv3r just in case this is rebased on top of master so i can merge without removing your verified tag?

**Comment by oliv3r** [comment]
> @oliv3r just in case this is rebased on top of master so i can merge without removing your verified tag?

Not sure what you meant, but it should have been rebased on master (as I always do a fetch; rebase; cycle), but since it's merged, I guess its good :)

**Comment by Ansuel** [comment]
(to give you some context... to flag pr as merged they needs to be merged with --ff-only so they have to be rebased on top of master... and by rebasing on top of master i would have dropped your verified tag as i would had to resign them for the force push... but they were rebased so np :D)
